warehouse/events/models.py

```diff
@@ -0,0 +1,61 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from sqlalchemy import Column, DateTime, ForeignKey, String, orm, sql
+from sqlalchemy.dialects.postgresql import JSONB, UUID
+from sqlalchemy.ext.declarative import AbstractConcreteBase, declared_attr
+
+from warehouse import db
+
+
+class Event(AbstractConcreteBase):
+    tag = Column(String, nullable=False)
+    time = Column(DateTime, nullable=False, server_default=sql.func.now())
+    ip_address = Column(String, nullable=False)
+    additional = Column(JSONB, nullable=True)
+
+
+class HasEvents:
+    @declared_attr
+    def events(cls):  # noqa: N805
+        cls.Event = type(
```

Comment:
My only concern with this, is that ``@declared_attr`` is designed to dynamically compute a declared attribute (which this is doing) but it's not really intended to *also* directly mutate the class. I think it will probably work fine?

However, It might be cleaner to do something like:

```python
class HasEvents:
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        cls.Event = ...
        return cls

    @declared_attr
    def events(cls):
         return orm.relationship(cls.Event, cascade="all, delete-orphan", lazy=True)
```

You might also move some of the dynanism up a level:


```python
class Event(AbstractConcreteBase):
    tag = Column(String, nullable=False)
    time = Column(DateTime, nullable=False, server_default=sql.func.now())
    ip_address = Column(String, nullable=False)
    additional = Column(JSONB, nullable=True)

    @declared_attr
    def __tablename__(cls):
       "_".join([cls.__name__.removesuffix("Event").lower(), "events"])

    @declared_attr
   def __mapper_args__(cls):
        return {'polymorphic_identity': cls.__name__,
                'concrete': True} if cls.__name__ != "Event" else {}


 
class HasEvents:
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        cls.Event = type(f"{cls.__name__}Event", (Event, db.Model), dict(source_id=..., source=...))
        return cls

    @declared_attr
    def events(cls):
         return orm.relationship(cls.Event, cascade="all, delete-orphan", lazy=True)

```

This is all completely untested, and you may be able to even do something like:

```python
class Event(AbstractConcreteBase):

    tag = Column(String, nullable=False)
    time = Column(DateTime, nullable=False, server_default=sql.func.now())
    ip_address = Column(String, nullable=False)
    additional = Column(JSONB, nullable=True)

    @declared_attr
    def __tablename__(cls):
       "_".join([cls.__name__.removesuffix("Event").lower(), "events"])

    @declared_attr
   def __mapper_args__(cls):
        return {'polymorphic_identity': cls.__name__,
                'concrete': True} if cls.__name__ != "Event" else {}

    @declared_attr
    def source_id(cls):
        return  source_id=Column(
                    UUID(as_uuid=True),
                    ForeignKey(
                        "%s.id" % cls._parent_class.__tablename__,
                        deferrable=True,
                        initially="DEFERRED",
                    ),
                    nullable=False,
                )
    @declared_attr
    def source(cls):
        orm.relationship(cls._parent_class)

    def __init_subclass__(cls, /, parent_class, **kwargs):
        super().__init_subclass__(cls, **kwargs)
        cls._parent_class = parent_class
        return cls


 
class HasEvents:
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        cls.Event = type(f"{cls.__name__}Event", (Event, db.Model), dict(source_id=..., source=...))
        return cls

    @declared_attr
    def events(cls):
         return orm.relationship(cls.Event, cascade="all, delete-orphan", lazy=True)
```

Feel free to ignore this of course, just feels a little easier to read and follow what's going on.

Reply:
I was able to get the 3rd option working: ef24030

(This also simplified the migration a bit)